core/app/models/spree/menu_item.rb

```diff
@@ -0,0 +1,84 @@
+module Spree
+  class MenuItem < Spree::Base
+    belongs_to :menu
+    belongs_to :linked_resource, polymorphic: true
+
+    acts_as_nested_set dependent: :destroy
+
+    before_save :reset_link_attributes, :build_path, :paremeterize_code
+
+    ITEM_TYPE = %w[Link Promotion Container]
+
+    LINKED_RESOURCE_TYPE = ['URL']
+    STATIC_RESOURCE_TYPE = ['Home Page']
+    DYNAMIC_RESOURCE_TYPE = ['Spree::Product', 'Spree::Taxon']
+
+    LINKED_RESOURCE_TYPE.unshift(*STATIC_RESOURCE_TYPE)
+    LINKED_RESOURCE_TYPE.push(*DYNAMIC_RESOURCE_TYPE)
+
+    validates :name, presence: true
+    validates :menu, presence: true
+    validates :item_type, inclusion: { in: ITEM_TYPE }
+    validates :linked_resource_type, inclusion: { in: LINKED_RESOURCE_TYPE }
+    validate :check_for_root, on: :create
+
+    has_one :menu_item_image, as: :viewable, dependent: :destroy, class_name: 'Spree::MenuItemImage'
```

Comment:
@MatthewKennedy can we skip the additional model here and just have `has_one_attached` ?

Reply:
@damianlegawiec I'm happy to change this back to `has_one_attached`, it is the simplest solution and the one I went with first, just a consideration before I go ahead.

Retrieving the image through the json api serializer, I only see this done using the external model way.

I can change this to use the existing `Spree::Image` model rather than `Spree::MenuItem::Image` model, and still get the image through the serializer for the API.

What do yo think is best?

Reply:
There is also a bug with awesome nested set and active storage

If you use `acts_as_nested_set` and `has_one_attached` in the same model the attachment is lost if you attach one to a newly created item.

The bug is not present when use use active storage via a separate model, in the current way.